### PR TITLE
Metric

### PR DESCRIPTION
**Pull request by kebinC**

### CODE REVIEW

File: fluid/PaddleCV/metric_learning/README.md [hunk 1]
```diff
@@ -17,146 +17,88 @@ Running sample code in this directory requires PaddelPaddle Fluid v0.14.0 and la
...
-|model | ResNet50 | SE-ResNeXt-50
+|pretrain model | softmax | arcmargin
 |- | - | -:
-|[triplet loss]() | 57.36% | 51.62%
-|[eml loss]() | 58.84% | 52.94%  
-|[quadruplet loss]() | 62.67% | 56.40%
+|[without fine-turned]() | 77.42% | 78.11%
+|[fine-turned with triplet]() | 78.37% | 79.21%
+|[fine-turned with quadruplet]() | 78.10% | 79.59%
+|[fine-turned with eml]() | 79.32% | 80.11%
```
Comment (qingqing01) [inline fluid/PaddleCV/metric_learning/README.md]: 如果不提供模型，把这里的()删掉吧

Comment (qingqing01) [reply fluid/PaddleCV/metric_learning/README.md]: 最后可以加下loss对应的参考文献吧，直接看名字，eml其实不懂是啥。

File: fluid/PaddleCV/metric_learning/losses/arcmarginloss.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import math
+import paddle.fluid as fluid
+
+class arcmarginloss():
```
Comment (qingqing01) [inline fluid/PaddleCV/metric_learning/losses/arcmarginloss.py]: class的命名规范，应该是大写吧，其他文件相同。

https://www.python.org/dev/peps/pep-0008/#class-names

File: fluid/PaddleCV/metric_learning/losses/arcmarginloss.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    def loss(self, input, label):
+        out = self.arc_margin_product(input, label, self.class_dim, self.margin, self.scale, self.easy_margin)
+        #loss = fluid.layers.softmax_with_cross_entropy(logits=out, label=label)
+        out = fluid.layers.softmax(input=out)
+        loss = fluid.layers.cross_entropy(input=out, label=label)
+        return loss, out
+
+    def arc_margin_product(self, input, label, out_dim, m, s, easy_margin=False):
+        #input = fluid.layers.l2_normalize(input, axis=1)
```
Comment (qingqing01) [inline fluid/PaddleCV/metric_learning/losses/arcmarginloss.py:23]: 现在换成l2_norm还会有问题吗？

File: fluid/PaddleCV/metric_learning/losses/commonfunc.py [hunk 4]
```diff
@@ -1,41 +1,15 @@
...
-    d = d + np.eye(len(fea)) * 1e8
-    sorted_index = np.argsort(d, 1)
-    res = 0
-    for i in range(len(fea)):
-        pred = lab[sorted_index[i][0]]
-        if lab[i] == pred:
-            res += 1.0
-    res = res/len(fea)
-    return res
+from __future__ import absolute_import
```
Comment (qingqing01) [inline fluid/PaddleCV/metric_learning/losses/commonfunc.py]: 文件命名可以是： utils.py?

File: fluid/PaddleCV/metric_learning/losses/commonfunc.py [hunk 5]
```diff
@@ -61,18 +35,17 @@ def calculate_order_dist_matrix(feature, batch_size, samples_each_class):
...
     a2 = fluid.layers.square(feature)
     a2 = fluid.layers.reduce_sum(a2, dim = 1)
+    #fluid.layers.Print(a2, summarize=10)
     d = fluid.layers.elementwise_add(-2*ab, a2, axis = 0)
     d = fluid.layers.elementwise_add(d, a2, axis = 1)
     d = fluid.layers.reshape(d, shape = [-1, 1])
     index = generate_index(batch_size, samples_each_class)
     index_var = fluid.layers.create_global_var(shape=[batch_size*batch_size], value=0, dtype='int32', persistable=True)
     index_var = fluid.layers.assign(index, index_var)
+    #index_var.stop_gradient = True
```
Comment (qingqing01) [inline fluid/PaddleCV/metric_learning/losses/commonfunc.py]: 去掉debug code，包括上面layers.Print?

File: fluid/PaddleCV/metric_learning/losses/arcmarginloss.py (hunk 3 above)
Comment (kebinC) [reply fluid/PaddleCV/metric_learning/losses/arcmarginloss.py:23]: 之前用的版本好像还有些问题，不知道现在什么情况。不过拆开来写跑的要稳定些。

File: fluid/PaddleCV/metric_learning/README.md (hunk 1 above)
Comment (kebinC) [reply fluid/PaddleCV/metric_learning/README.md]: done

File: fluid/PaddleCV/metric_learning/losses/arcmarginloss.py (hunk 2 above)
Comment (kebinC) [reply fluid/PaddleCV/metric_learning/losses/arcmarginloss.py]: done

File: fluid/PaddleCV/metric_learning/losses/commonfunc.py (hunk 5 above)
Comment (kebinC) [reply fluid/PaddleCV/metric_learning/losses/commonfunc.py]: done

File: fluid/PaddleCV/metric_learning/README.md (hunk 1 above)
Comment (qingqing01) [reply fluid/PaddleCV/metric_learning/README.md]: 文档里还是没加参考文献吧？

**Review by qingqing01 — APPROVED** [review]

**Comment by wanghaoshuang** [comment]
用用户提问：arcmargin loss的实现里，为什么不用l2_normalize这个接口，而要自己写？
https://github.com/PaddlePaddle/models/blob/develop/PaddleCV/metric_learning/losses/arcmarginloss.py#L46

**Comment by kebinC** [comment]
> 用用户提问：arcmargin loss的实现里，为什么不用l2_normalize这个接口，而要自己写？
> https://github.com/PaddlePaddle/models/blob/develop/PaddleCV/metric_learning/losses/arcmarginloss.py#L46

 当时实现的时候，l2_normalize 这个接口有bug，训练出nan，现在应该没问题了，可以改过来。
